packages/engine/paima-sm/src/index.ts

```diff
@@ -51,10 +52,39 @@ import type {
 import { ConfigNetworkType } from '@paima/utils';
 import assertNever from 'assert-never';
 import { keccak_256 } from 'js-sha3';
+import stableStringify from 'json-stable-stringify';
 
 export * from './types.js';
 export type * from './types.js';
 
+function blockHash(
+  chainData: ChainData,
+  prevBlockHash: null | string,
+  successfulTxs: string[],
+  failedTxs: string[]
+): {
+  header: PostExecutionBlockHeader;
+  hash: string;
+} {
+  const header = {
+    version: 1 as const,
+    prevBlockHash,
+    mainChainBlochHash: strip0x(chainData.blockHash),
+    blockHeight: chainData.blockNumber,
+    msTimestamp: chainData.timestamp * 1000,
+    successTxsHash: keccak_256(successfulTxs.join('|')),
+    failedTxsHash: keccak_256(failedTxs.join('|')),
+  };
+  return {
+    header,
+    hash: keccak_256(stableStringify(header)),
```

Comment:
out of all the decisions, using `stableStringify` here feels like the most questionable